### DISCUSSION TITLE
Shape Idea

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#5798] [Language Ideas]
I was thinking about shapes a bit and had an idea I wanted to share.

This concept is inspired by records and required properties.

Minimal usage and example:
```
public shape IName {
   string Name {get; set;}
}

public class Person {
  public string Name {get; set;}
}

public class Location {
  public string DisplayName {get; set;}
}

var PS = person as IName;
var LS = location as IName with {
  Name = DisplayName,
};
var OS = object as IName with {
  get_Name = ()=> string.Empty,
  set_Name = value => { },
};
```

1.  The ```shape``` keyword means you're declaring an ```interface``` that can be used as a shape.
2.  When ```as IShape``` appears, the runtime will generate a "builder" type that has init-only (no getter) Action/Func properties that correspond to all matching members of the shape.  All members are automatically assigned to a member of the matching name and type from the source object and then the object is "built".  When the object is built, an anonymous type is emitted that implements the interface (shape).  It is an error for all members not to be mapped.  This may require the language team to relax the restriction on having multiple properties with the same name.  This should not be an issue though since all properties are set-only and will never have their value retrieved.
3.  When the code ```as IShape with``` appears, the developer can override automatically inferred mappings and provide mappings where one could not be inferred.  The object is not built until the end of the ```with``` block. 

Other thoughts or challenges:
This would require that multiple properties with the same name (although different return type) be allowed.  For example:
```
public shape Example1 {
  void DoWork();
  void DoWork(int Id);
}
```

Might have the mappings manually provided like so:
```
var E1 = O as Example1 with {
  DoWork = () => { },
  DoWork = Id => { },
};
```

The anonymous object should contain an internal ```Target``` which contains the object being shaped.  This way other comparisons on the shape can be done the Target and not the shape instance.

Maybe ```as``` isn't the right word since it already has meaning.  Maybe a new word, ```to``` should be introduced?

Maybe an unmapped member should default to giving a compile time warning and throwing a ```NotImplementedException``` instead?

**Comment by HaloFour** [comment]
This looks more like a discussion around anonymous implementation of interfaces than shapes.  The existing proposal doesn't differentiate between shapes and interfaces, and the work to support static members in interfaces is part of that effort.
